Additional Kext Tests

jeschu1 opened the pull request:
Added Tests for 
- TryGetFileIsFlaggedAsInRoot
- ShouldHandleVnodeOpEvent
- VirtualizationRoot_VnodeIsOnAllowedFilesystem
- VirtualizationRoot_IsValidRootHandle

jeschu1 commented on ProjFS.Mac/PrjFSKextTests/KauthHandlerTests.mm:
```diff
@@ -1,10 +1,33 @@
+{
+    buf = procname;
+}
+
```
I don't love having the definitions and 3 functions here, but i'm reluctant to move them unless we think we'll use them elsewhere.  
If folks think we should move please also suggest a place/filename.

jeschu1 replied:
> The thing to bear in mind with having them defined here is that they equally apply to other test suites because they're all linked into a single bundle.

Thats a good point to remember.  Moving to MockProc.  We'll get more specific / move in the future if needbe.

jeschu1 commented on ProjFS.Mac/PrjFSKextTests/KauthHandlerTests.mm:
```diff
@@ -1,10 +1,33 @@
 
 using std::shared_ptr;
 
+#define KAUTH_RESULT_ALLOW    (1)
```
These could also be enums.  Not sure what we gain/lose by one versus another.  If we do make it an enum note to self to initialize at 1.

pmj replied:
It doesn't really matter either way, except if we move it to a header, you might need to surround the definitions with `#ifdef KAUTH_RESULT_ALLOW` to avoid clashing with the kernel headers when building the testable kext.

jeschu1 replied:
I switched them to enums so we don't have to worry about clashes

jeschu1 commented on ProjFS.Mac/PrjFSKextTests/KauthHandlerTests.mm:
```diff
@@ -58,4 +81,179 @@ - (void)testShouldIgnoreVnodeType {
+
+    
+    // Test Success Case
+    XCTAssertTrue(ShouldHandleVnodeOpEvent(
```
I tested evey category of failure for this function.  Side note it is super cool that I can do that :-).  Big thanks to @pmj for helping us get this far :-D.

We may decide to test every scenario in every category at a later date.

pmj commented on ProjFS.Mac/PrjFSKextTests/KauthHandlerTests.mm:
```diff
@@ -58,4 +81,179 @@ - (void)testShouldIgnoreVnodeType {
+    XCTAssertTrue(TryGetFileIsFlaggedAsInRoot(testVnode.get(), NULL, &fileFlaggedInRoot));
+    XCTAssertFalse(fileFlaggedInRoot);
+    
+    testVnode->values.getattr = 100;
```
Given this is a bit field, a hex literal might be more appropriate.

pmj commented on ProjFS.Mac/PrjFSKextTests/KauthHandlerTests.mm:
```diff
@@ -58,4 +81,179 @@ - (void)testShouldIgnoreVnodeType {
+    
+    // Test Success Case
+    XCTAssertTrue(ShouldHandleVnodeOpEvent(
+                                            &perfTracer,
```
Nit: the indentation is a bit much here, 1 level ought to be enough.

pmj commented on ProjFS.Mac/PrjFSKextTests/KauthHandlerTests.mm:
```diff
@@ -58,4 +81,179 @@ - (void)testShouldIgnoreVnodeType {
+    XCTAssertEqual(kauthResult, KAUTH_RESULT_DEFER);
+
+    
+    // Test invalid File System
```
What's invalid about this file system?

jeschu1 replied:
"hfs"

pmj replied:
That's [one of the 2 supported file systems](https://github.com/Microsoft/VFSForGit/blob/98acb8c7eaf92b26d514f68bf3f9dc88095ad475/ProjFS.Mac/PrjFSKext/VirtualizationRoots.cpp#L596) though?

FWIW, this test fails when I run it locally.

jeschu1 replied:
Thanks for that, you are correct here.  Updating.

pmj commented on ProjFS.Mac/PrjFSKextTests/KauthHandlerTests.mm:
```diff
@@ -58,4 +81,179 @@ - (void)testShouldIgnoreVnodeType {
+    vtype vnodeType;
+    uint32_t vnodeFileFlags;
+    int pid;
+    char procname[MAXCOMLEN + 1] = "";
```
This shadows the global variable of the same name, is that intentional?

jeschu1 replied:
No it wasn't, fixing..

pmj commented on ProjFS.Mac/PrjFSKextTests/KauthHandlerTests.mm:
```diff
@@ -1,10 +1,33 @@
+char * procname;
+void proc_name(int pid, char * buf, int size)
+{
+    buf = procname;
```
This essentially a no-op, `buf` is passed by value. You'll need to use `strlcpy()` to write to the supplied buffer. (But then `procname` is never initialised; is this function called by the current tests? If not, perhaps just put an `assert(false)` in here until we actually need to implement it properly.)

pmj commented on ProjFS.Mac/PrjFSKextTests/KauthHandlerTests.mm:
```diff
@@ -58,4 +81,179 @@ - (void)testShouldIgnoreVnodeType {
+
+    
+    // Test with finder trying to populate an empty file
+    strcpy(procname, "finder");
```
Nit: The `"Finder"` process is capitalised, but it shouldn't matter for the purposes of the test.

jeschu1 commented on ProjFS.Mac/PrjFSKext/KauthHandlerTestable.hpp:
```diff
@@ -12,5 +13,20 @@
 KEXT_STATIC bool IsFileSystemCrawler(const char* procname);
 KEXT_STATIC bool ShouldIgnoreVnodeType(vtype vnodeType, vnode_t vnode);
+KEXT_STATIC bool ShouldHandleVnodeOpEvent(
+                                          // In params:
```
Thanks for bringing this up.

When I autoformat in XCode (Cntrl-I), it does this level of indenting.
I really don't like fighting autoformatting of editors in new code.

Am I correct that this is the autoformat of XCode?  If not, please advise me on how to fix.  If so, can we consider moving to autoformat standards for new code?  @wilbaker

pmj replied:
Xcode's default indentation rules are unfortunately optimised for Objective-C's named function parameters and tend to cause excessive indentation for C and C++ style function calls. I personally have "Automatically indent based on syntax" disabled (Preferences -> Text Editing -> Indentation); as far as I'm aware, this is the only way to turn off the crazy levels of function argument indentation. Unfortunately, unlike indentation level rules (spaces/tabs, number of spaces) this isn't something you can set on a per-file or per-project basis.

I don't think that level of indentation is in any way helpful, so I'm not keen on changing our style guidelines simply because Xcode isn't designed particularly well for C or C++ editing.

jeschu1 commented on ProjFS.Mac/PrjFSKextTests/MockVnodeAndMount.cpp:
```diff
@@ -104,6 +99,12 @@ void vnode_putname(const char* name)
 
+int vnode_getattr(vnode_t vp, struct vnode_attr *vap, vfs_context_t ctx)
+{
+    vap->va_flags = vp->values.getattr;
```
Great tips here thank you!

jeschu1 commented on ProjFS.Mac/PrjFSKextTests/MockVnodeAndMount.hpp:
```diff
@@ -18,14 +18,25 @@ struct mount
+
+struct VnodeMockValues
+{
+    int getattr = 0;
```
Definitely prefer this as a cleaner approach.  It was not really an int :-), uint32_t as you may have guessed.

I updated for va_flags.

Moving some of the existing fields to attrValues, seems like a meatier change (or i'm missing something).  Can you elaborate?

pmj replied:
> Moving some of the existing fields to attrValues, seems like a meatier change (or i'm missing something). Can you elaborate?

Yes, that'd be a bit bigger. We can fix that up on demand, no need to lump it in with this change. The new version looks good to me!

jeschu1 commented on ProjFS.Mac/PrjFSKextTests/KauthHandlerTests.mm:
```diff
@@ -1,10 +1,17 @@
 
 using std::shared_ptr;
 
+proc_t vfs_context_proc(vfs_context_t ctx)
```
@pmj this was annoying but when I move this function to MockProc, the linker complained about a dual definition.  Do you see any obvious problems here?

pmj replied:
Hmm, I don't know what might have been the problem there. I just tried it and it worked straight away:

https://github.com/pmj/VFSForGit/commit/98d80aa41262d47efcb05ffa700f0552951c1675

jeschu1 replied:
Thanks for trying that @pmj, yeah it seems totally fine now.  Perhaps I was working too late :)

jeschu1 commented on ProjFS.Mac/PrjFSKextTests/KauthHandlerTests.mm:
```diff
@@ -58,4 +65,189 @@ - (void)testShouldIgnoreVnodeType {
+            &vnodeType,
+            &vnodeFileFlags,
+            &pid,
+            procname,
```
How vigelent should we be abount freeing memory?
We aren't freeing procname here, but the memory will be returned when the test is over correct?

pmj replied:
`procname` is a local variable of array type, so it's allocated on the stack and automatically disappears when the function returns, so in this case there's nothing to worry about.

jeschu1 replied:
I copied the pattern you had for Path.  Using string is a nice way to not have to worry about free-ing.

pmj commented on ProjFS.Mac/PrjFSKextTests/MockProc.cpp:
```diff
@@ -0,0 +1,23 @@
+
+void SetProcName(char* procName)
+{
+    return_proc_name = strdup(procName);
```
We should `free()` the previous memory before overwriting the pointer. Or just use a `std::string` for the variable holding the mock value and don't worry about memory.

pmj commented on ProjFS.Mac/PrjFSKextTests/MockProc.cpp:
```diff
@@ -0,0 +1,23 @@
+    }
+}
+
+void SetProcName(char* procName)
```
This parameter should be a `const char*` so we can use literals.

pmj commented on ProjFS.Mac/PrjFSKextTests/MockProc.cpp:
```diff
@@ -0,0 +1,23 @@
+    return 1;
+}
+
+char* return_proc_name;
```
I don't think we need this variable outside this compilation unit, so it can be `static`. I'm not sure what our naming convention for global variables is, but for file-level `static` variables it's `s_camelCase`.

pmj commented on ProjFS.Mac/PrjFSKextTests/MockProc.cpp:
```diff
@@ -0,0 +1,23 @@
+{
+    if (return_proc_name != nullptr)
+    {
+        strcpy(buf, return_proc_name);
```
Please always use `strlcpy`, or in some exceptional circumstances `strncpy`. `strcpy` is almost never appropriate, as it doesn't check for enough space in the destination.

[The real `proc_name` implementation](https://github.com/apple/darwin-xnu/blob/a449c6a3b8014d9406c2ddbdc81795da24aa7443/bsd/kern/kern_proc.c#L874) uses `strlcpy`, so let's definitely use that please.

jeschu1 commented:
@pmj thanks for the quick turnaround and pointers here.  My C is rusty for sure, so I definitely appreciate the tips :-).  I addressed your comments in a separate commit for easy review.  I'll squash before merging.

pmj commented on ProjFS.Mac/PrjFSKextTests/MockProc.cpp:
```diff
@@ -1,23 +1,24 @@
-    {
-        strcpy(buf, return_proc_name);
-    }
+    strlcpy(buf, s_procName.c_str(), MIN(size, MAXPATHLEN));
```
Checking for `MAXPATHLEN` isn't really appropriate here. There is a `MAXCOMLEN` limitation in the kernel, but for our purposes I think we can ignore that for the sake of simplicity:

```
strlcpy(buf, s_procName.c_str(), size);
```

jeschu1 replied:
Sigh yes :-), copy/paste error

wilbaker commented on ProjFS.Mac/PrjFSKextTests/KauthHandlerTests.mm:
```diff
@@ -58,4 +60,187 @@ - (void)testShouldIgnoreVnodeType {
+    XCTAssertTrue(TryGetFileIsFlaggedAsInRoot(testVnode.get(), NULL, &fileFlaggedInRoot));
+    XCTAssertFalse(fileFlaggedInRoot);
+    
+    testVnode->attrValues.va_flags = 0x00000100;
```
Does `0x00000100` have special meaning? Or was it just something other than `FileFlags_IsInVirtualizationRoot`?

jeschu1 replied:
Just something different

wilbaker commented on ProjFS.Mac/PrjFSKextTests/KauthHandlerTests.mm:
```diff
@@ -58,4 +60,187 @@ - (void)testShouldIgnoreVnodeType {
+
+    testVnode->errors.getattr = EBADF;
+    XCTAssertFalse(TryGetFileIsFlaggedAsInRoot(testVnode.get(), NULL, &fileFlaggedInRoot));
+}
```
Consider also adding a test that sets `va_flags` to `FileFlags_IsInVirtualizationRoot` OR-ed (`|`) with something else and assert that `fileFlaggedInRoot` is true.

pmj replied:
FWIW, possible OS-defined values for flags are:

 * `UF_NODUMP`
 * `UF_IMMUTABLE`
 * `UF_APPEND`
 * `UF_OPAQUE`
 * `UF_HIDDEN`
 * `SF_ARCHIVED`
 * `SF_IMMUTABLE`
 * `SF_APPEND`

See the manpage for `chflags(2)`. These values should be defined in `<sys/stat.h>`.

wilbaker commented on ProjFS.Mac/PrjFSKextTests/KauthHandlerTests.mm:
```diff
@@ -58,4 +60,187 @@ - (void)testShouldIgnoreVnodeType {
+    shared_ptr<vnode> testVnode = vnode::Create(testMount, "/foo");
+    testVnode->attrValues.va_flags = FileFlags_IsInVirtualizationRoot;
+    PerfTracer perfTracer;
+    vfs_context_t _Nonnull context = vfs_context_create(NULL);
```
Nit: `NULL` -> `nullptr`

wilbaker commented on ProjFS.Mac/PrjFSKextTests/KauthHandlerTests.mm:
```diff
@@ -58,4 +60,187 @@ - (void)testShouldIgnoreVnodeType {
+    XCTAssertEqual(kauthResult, KAUTH_RESULT_DEFER);
+    
+    
+    // Test Invalid Action Bit
```
Minor:  When I read "Invalid Action Bit" it sounded like we'd be passing in a bit that doesn't correspond to a well known `KAUTH_VNODE_` value.

Consider updating this comment to something like:

`KAUTH_VNODE_ACCESS should not be handled`

wilbaker commented on ProjFS.Mac/PrjFSKextTests/KauthHandlerTests.mm:
```diff
@@ -58,4 +60,187 @@ - (void)testShouldIgnoreVnodeType {
+            &perfTracer,
+            context,
+            testVnode.get(),
+            KAUTH_VNODE_ACCESS,
```
Also add a test for `KAUTH_VNODE_ACCESS` OR-ed with something we do otherwise care about?

e.g. `KAUTH_VNODE_ACCESS | KAUTH_VNODE_DELETE`

jeschu1 commented on ProjFS.Mac/PrjFSKextTests/MockProc.hpp:
```diff
@@ -0,0 +1,14 @@
+extern "C" void proc_name(int pid, char * buf, int size);
+extern "C" proc_t vfs_context_proc(vfs_context_t ctx);
+
+void SetProcName(const char* procName);
```
Forgive my rusty C++, but if we don that can we call "SetProcName("Finder")"?

wilbaker replied:
Yep, if we make the parameter `const std::string&` we can still call the function with `char*`.

This is allowed because `std::string` has a [converting constructor](https://stackoverflow.com/questions/15077466/what-is-a-converting-constructor-in-c-what-is-it-for) for [char*](http://www.cplusplus.com/reference/string/string/string/).  A temporary `std::string` will be automatically constructed for us from the `char*`.

pmj replied:
Yeah, `const std::string&` is OK here. It's definitely less efficient, but we don't care about that much in the tests, and allows us to pass either `const char*` or `std::string` objects in.

jeschu1 commented on ProjFS.Mac/PrjFSKextTests/MockVnodeAndMount.hpp:
```diff
@@ -72,6 +72,12 @@ struct mount
+struct VnodeMockErrors
+{
+    errno_t getpath = 0;
+    errno_t getattr = 0;
```
VnodeMockErrors (Struct Name) signifies they are errors for the listed function calls.
With the struct name in mind, the error seems a bit redundant.
All that would be in this class are return codes for the specified function names.

pmj replied:
I agree with @jeschu1 here; actual uses of this struct in test code (e.g. elsewhere in this PR) currently look like this:

```
testVnode->errors.getattr = EBADF;
```

That makes perfect sense to me, including sticking with the capitalisation of function `vnode_getattr` vs. the camelCase convention we'd be applying if we were coming up with the name from scratch. I don't think the alternatives,

```
testVnode->errors.getAttrError = EBADF;
// or
testVnode->errors.getAttrResult = EBADF;
```

are an improvement.

If we're going to change anything here, the variant I'd be open to is:

```
testVnode->errors.vnode_getattr = EBADF;
```

I like this despite it flagrantly disregarding our naming conventions - it makes it even more obvious we're providing the mock error code for `vnode_getattr()`.

wilbaker replied:
@jeschu1 @pmj that sounds good to me, thanks for the detailed explanations!

pmj commented on ProjFS.Mac/PrjFSKext/KauthHandlerTestable.hpp:
```diff
@@ -12,5 +13,20 @@
 
 KEXT_STATIC_INLINE bool FileFlagsBitIsSet(uint32_t fileFlags, uint32_t bit);
 KEXT_STATIC_INLINE bool ActionBitIsSet(kauth_action_t action, kauth_action_t mask);
+KEXT_STATIC_INLINE bool TryGetFileIsFlaggedAsInRoot(vnode_t vnode, vfs_context_t context, bool* flaggedInRoot);
```
The mock KPIs currently don't use the context pointer for anything, but we could change that to check it gets passed through correctly and consistently. (some APIs silently accept `nullptr` contexts in some circumstances while KPing in others, for example)

So I'd agree, we should probably try to write our tests consistently. Where we don't really care about vfs contexts, we can just create one in `setUp` and destroy it in `tearDown` and use it in each of the test methods. That's not significantly more work than passing `nullptr`.

jeschu1 replied:
@wilbaker since KauthHandlerTestable is a header file if we add _Nonnull for one paramter we must add it for every parameter.  

KauthHandler.cpp does not have this requirement since it isn't a header.

So to get the signatures to match we'd either need to
1. Add a nonnullable/nullable tag to every param in Kauthhandler.cpp which creates a lot of noise.  Don't do that.
2. Use " #pragma clang assume_nullable begin/end" which @pmj has advised against.

For now I'll complete this PR without adding `_Nonnull `but we can pick the discussion up at a later time.

wilbaker replied:
> @wilbaker since KauthHandlerTestable is a header file if we add _Nonnull for one paramter we must add it for every parameter.

> KauthHandler.cpp does not have this requirement since it isn't a header.

Ahh I didn't realize there was a difference in the requirements for these specifiers (between headers and source).

> Add a nonnullable/nullable tag to every param in Kauthhandler.cpp which creates a lot of noise. Don't do that.

I actually think tagging all of the pointer params isn't such a bad idea.  It would get us some additional protection from the compiler and help convey the expectations for those parameters.  

However, I agree there's no need to clean that up now (or block this PR on it).

wilbaker replied:
> Ahh I didn't realize there was a difference in the requirements for these specifiers (between headers and source).

This would also explain why I didn't hit this issue with the vnode cache, I believe I specified the nullability for all pointer parameters.
